### DISCUSSION TITLE
wasmCloud Q4 2024 Roadmap

### DISCUSSION THREAD

**Discussion by brooksmtownsend** [wasmCloud/wasmCloud#3119] [Community RFC]
Hello all!

For the past year, we've published roadmaps for wasmCloud after community planning sessions during our weekly Wednesday community meeting. https://wasmcloud.com/docs/roadmap/ We're coming up on the final quarter of this year, and we're looking forward to our next planning meeting!

This discussion is for suggesting features, prioritizing fixes, and general brainstorming for the priorities of wasmCloud maintainers + contributors going into Q4. Our previous roadmap focused on community contributions, improving developer workflows, and finalizing some support for WASI interfaces. Based on the discussion here, we'll determine our next roadmap's focus.

Please feel free to comment here whether you're a long-time wasmCloud user, maintainer, first-time contributor, or just curious onlooker that noticed something missing!

**Comment by brooksmtownsend** [comment]
I think https://github.com/wasmCloud/wasmCloud/issues/3120 is a core piece of our next quarter of work for the future of wasmCloud 😄

**Comment by ossfellow** [comment]
This will mark the beginning of the third quarter that I have joined the wasmCloud community, and decided to build a complex decision engine on top of wasmCloud platform.
And as I've been learning my way through the inner plumbing of the platform, and how the community functions, I've come to the conclusion that the best way to balance my needs and wants of the platform, with plans and objectives of the maintainers, and the demands of a growing community, is to contribute additive features, and bug fixes.

The maintainers have been very responsive to suggestions about communication improvements, and I've witnessed significant improvements in documentation and details provided about the changes made to the platform design and its APIs.
However, still, writing a provider can sometimes feel like a trap shooting exercise, where a change in the _provider-sdk_, or the _host_, or even _wRPC transport_, can render the _under-construction_ provider _non-functional_ and suddenly _incompatible_.

This issue will only grow as the wasmCloud community and ecosystem grows; so, sooner or later, we - this community - need a solution, which could accommodate the _growth_ and _evolution_ of the _core ABIs and APIs_ of the platform, with _core API stability_, _good documentation_, and _kept-informed attitude_  needed by the _core service layer_ contributors - as a positive force in platform's growth.
Fortunately, we already have a battle tested solution in the form of `wasmCloud RFCs`, which, if expanded to include all planned changes to the _core ABIs and APIs_ of the platform, can fill a major communication gap , and let contributors sync their work with the work that's been done by the maintainers.

In Q4, I will be contributing the `nats-provider-blobstore` (`wasi:blobstore@0.2.0-draft` proposal). I explored implementing the WASI Blobstore interface spec - as opposed to the `wrpc:blobstore@0.2.0` - aided by the codebase of the wasmCloud `runtime` helper functions (`keyvalue.rs`, etc.), but due to concern about additional maintenance efforts, I've decided to abandon the idea, and implement the _wrpc_ variant of the spec.
However, if we want to slowly move away from the current form of _polyfilling_, or _wrpcfication_ of WASI interface proposals, this could be an opportunity for an experimentation about any missing pieces (i.e. public APIs) needed to directly implement the WASI specs.

I also want to contribute a new page to the _Kubernetes_ section documenting a `secure stack` ( i.e. armed) for running wasmCloud and its requisite dependencies (NATS and Wadm); but I can only start working on this close to the holidays, and there's a small chance that the completion of the work could slip into early Q1 2025.

Lastly, on the top of my wish list remains seeing an _custom provider example_ just like the [custom-template](https://github.com/wasmCloud/wasmCloud/tree/main/examples/rust/providers/custom-template), but handling a `resource`, which lives outside wasmCloud (e.g. a file).
Also, I wish to see the same set of APIs leveraged by the _host_ and _providers_ for managing the lifecycle of _resources_.
These will enable and empower willing contributors like me to add custom (i.e. non-WASI) capabilities to the wasmCloud platform, which will ultimately increase its usability and appeal.

And we're just beginning...

**Comment by brooksmtownsend** [comment]
https://github.com/orgs/wasmCloud/projects/7/views/12
